bloptools/bayesian/__init__.py

```diff
@@ -0,0 +1,567 @@
+
+    def ask(
+        self,
+        tasks=None,
+        classifier=None,
+        strategy=None,
+        greedy=True,
+        n=1,
+        disappointment=0,
+        route=True,
+        cost_model=None,
+        n_test=1024,
+        optimize=True,
+    ):
+        """
+        Recommends the next $n$ points to sample.
+        """
+
+        if route:
+            unrouted_points = self.ask(
+                tasks=tasks,
+                classifier=classifier,
+                strategy=strategy,
+                greedy=greedy,
+                n=n,
+                disappointment=disappointment,
+                route=False,
+                cost_model=cost_model,
+                n_test=n_test,
+            )
+
+            routing_index, _ = utils.get_routing(self.current_inputs, unrouted_points)
+            return unrouted_points[routing_index]
+
+        if strategy.lower() == "quasi-random":
+            return self.unnormalize_inputs(self.sampler(n=n))
+
+        if not self._initialized:
+            raise RuntimeError("The agent is not initialized!")
+
+        if tasks is None:
+            tasks = self.tasks
+
+        if classifier is None:
+            classifier = self.classifier
+
+        if (not greedy) or (n == 1):
+            acqf = None
+
+            if strategy.lower() == "esti":  # maximize the expected improvement
+                acqf = acquisition.log_expected_sum_of_tasks_improvement
+
+            if acqf is None:
+                raise ValueError(f'Unrecognized strategy "{strategy}".')
+
+            X, loss = self.sample_acqf(acqf, optimize=optimize)
+
+            return self.unnormalize_inputs(np.atleast_2d(X))
+
+        if greedy and (n > 1):
+            dummy_tasks = [copy.deepcopy(task) for task in self.tasks]
+
+            inputs_to_sample = np.zeros((0, self.n_dof))
+
+            for i in range(n):
+                new_input = self.ask(strategy=strategy, tasks=dummy_tasks, classifier=classifier, greedy=True, n=1)
+
+                inputs_to_sample = np.r_[inputs_to_sample, new_input]
+
+                if not (i + 1 == n):  # no point if we're on the last iteration
+                    fantasy_multimodel = botorch.models.model.ModelList(*[task.regressor for task in self.tasks])
+                    fantasy_posterior = fantasy_multimodel.posterior(
+                        torch.tensor(self.normalize_inputs(new_input)).double()
+                    )
+
+                    fantasy_targets = fantasy_posterior.mean - disappointment * fantasy_posterior.variance.sqrt()
+                    new_targets = self.unnormalize_targets(fantasy_targets.detach().numpy())
+
+                    self.tell(new_inputs=new_input, new_targets=new_targets)
+
+            self.forget(n=n)  # forget what you saw here
+
+            return inputs_to_sample
+
+    def learn(
+        self, strategy, n_iter=1, n_per_iter=1, reuse_hypers=True, upsample=1, verbose=True, plots=[], **kwargs
+    ):
```

Comment:
Need a doc string for this method too as it's one of the most important methods.